### ISSUE TITLE
Add in MutationObserver

### ISSUE THREAD

**Issue by Pauan** [koute/stdweb#85]
I've been working on adding in [MutationObserver](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver), but I have some questions:

* When calling the `observe` method, you can pass in an `attributeFilter` attribute, which in JavaScript is an Array of strings. What should the type be in Rust? `Vec<String>`?

   If the correct type is `Vec<String>`, how do I convert it into a JavaScript Array so I can pass it to the `observe` method?

* `MutationObserver` accepts a callback. I took a look at how event listeners are implemented, and the callbacks have a `'static` trait bound.

   I assume that exists to guarantee that the callback will live indefinitely, because Rust doesn't have a garbage collector. Is that correct?

* I want to provide a nice Rust-y API for `MutationRecord`, something like this:

   ```rust
   pub enum MutationRecord {
       Attribute {
           target: Node,
           attribute_name: Option< String >,
           attribute_namespace: Option< String >,
           old_value: Option< String >,
       },

       CharacterData {
           target: Node,
           old_value: Option< String >,
       },

       ChildList {
           target: Node,
           added_nodes: NodeList,
           removed_nodes: NodeList,
           previous_sibling: Option< Node >,
           next_sibling: Option< Node >,
       },
   }
   ```

   I tried to add in an `impl TryFrom< Value > for MutationRecord` implementation, so I could convert from JS objects into `MutationRecord`, but I'm getting this error:
   ```
   error[E0277]: the trait bound `std::option::Option<webapi::node::Node>: webcore::try_from::TryFrom<webcore::value::Value>` is not satisfied
      --> src/webapi/mutation_observer.rs:124:71
       |
   124 |                         previous_sibling: js!( @{r}.previousSibling ).try_into()?,
       |                                                                       ^^^^^^^^ the trait `webcore::try_from::TryFrom<webcore::value::Value>` is not implemented for `std::option::Option<webapi::node::Node>`
       |
       = help: the following implementations were found:
                 <std::option::Option<bool> as webcore::try_from::TryFrom<webcore::value::Value>>
                 <std::option::Option<u8> as webcore::try_from::TryFrom<webcore::value::Value>>
                 <std::option::Option<u16> as webcore::try_from::TryFrom<webcore::value::Value>>
                 <std::option::Option<u32> as webcore::try_from::TryFrom<webcore::value::Value>>
               and 14 others
       = note: required because of the requirements on the impl of `webcore::try_from::TryInto<std::option::Option<webapi::node::Node>>` for `webcore::value::Value`
   ```

   Basically, I'm trying to convert from a JavaScript value (which is either `null` or a DOM `Node`) into an `Option< Node >` in Rust, but it's not working. How do I fix this?

You can see my (unfinished) work [here](https://github.com/Pauan/stdweb/blob/f77682bf4e4abcbe6b37bd341468815fffbf59db/src/webapi/mutation_observer.rs).

**Comment by koute** [comment]
> When calling the observe method, you can pass in an attributeFilter attribute, which in JavaScript is an Array of strings. What should the type be in Rust? Vec<String>?

If you're passing it into the JS then probably `&[&str]`. (Although it would be nice to support both `&[&str]` and `&[String]`, I guess, but that's not always easy to do due to Rust's trait system limitations.)

> I assume that exists to guarantee that the callback will live indefinitely, because Rust doesn't have a garbage collector. Is that correct?

That's correct. The callback can be called or deallocated at any time after it's passed into JS, so it can't hold any references to anything local. If one wants to share data with the insides of the callback either a global variable has to be used, or a reference counted one (with `Rc`).

> Basically, I'm trying to convert from a JavaScript value (which is either null or a DOM Node) into an Option< Node > in Rust, but it's not working. How do I fix this?

You don't, I already fixed it for you. (:

The lack of this conversion was an ommision, which I've just rectified on the newest `master`. It should work now.

**Comment by Pauan** [comment]
> If you're passing it into the JS then probably `&[&str]`.

That doesn't quite seem right to me. The API is called like this:

```rust
observer.observe( some_dom_node, MutationObserverInit {
    child_list: true,
    attribute: false,
    character_data: true,
    subtree: true,
    attribute_old_value: false,
    character_data_old_value: false,
    attribute_filter: vec![ "foo", "bar", "qux" ],
} );
```

In other words, it's not really intended for you to continue using `attribute_filter` after passing it into `observe`: its sole purpose is to pass into JS. So wouldn't `String` make more sense than `&str`?

Or are you suggesting `&str` because of the rare chance that somebody actually *does* want to use the `Vec` after passing it into `observe`?

In any case, regardless of whether I use `Vec<String>` or `&[&str]`, one question remains: how do I convert it into a JS Array of strings? Do I use `std::convert::From::from`?

----

> You don't, I already fixed it for you. (:

Great, thanks!

**Comment by koute** [comment]
> In other words, it's not really intended for you to continue using attribute_filter after passing it into observe: its sole purpose is to pass into JS. So wouldn't String make more sense than &str?

You have it backwards. (:

If a function doesn't need the ownership and only needs the data temporarily (e.g. during a single function call) you want it to take a borrow, that is - to take an `&[T]` instead of `Vec<T>` and `&str` instead of `String`. This has some practical implications:

1. `Vec<T>` and `String` require a heap allocation to construct, `&[T]` and `&str` don't. So in cases where you're **not** generating them dynamically by having the function take `&[T]` and `&str` you're not paying extra since you don't need to `.clone()` your `&[T]` before the call.
2. `Vec<T>` and `String` can be easily converted with a [deref coercion](https://doc.rust-lang.org/book/first-edition/deref-coercions.html) to `&[T]` and `&str` respectively by putting an ampersand before the variable name anyway.
3. In this particular case those variables **must** be copied anyway to expose them to JS so you don't gain anything by taking their ownership.

> In any case, regardless of whether I use Vec<String> or &[&str], one question remains: how do I convert it into a JS Array of strings? Do I use std::convert::From::from?

You should be able to call `.into()` on it to convert it to `Value`. (As an array of strings should always be convertible to a `Value` through [this impl](https://docs.rs/stdweb/*/stdweb/enum.Value.html#impl-From%3C&'a%20[T]%3E))

**Comment by Pauan** [comment]
> If a function doesn't need the ownership and only needs the data temporarily (e.g. during a single function call) you want it to take a borrow

Ah, I see what you mean. If it takes `&[&str]` then it's possible to do this:

```rust
observer.observe( some_dom_node, MutationObserverInit {
    child_list: true,
    attribute: false,
    character_data: true,
    subtree: true,
    attribute_old_value: false,
    character_data_old_value: false,
    attribute_filter: &[ "foo", "bar", "qux" ],
} );
```

This doesn't require a heap allocation (as opposed to `Vec` which does). You could even store the array in a `const`

----

> You should be able to call `.into()` on it to convert it to `Value`.

Thanks! It wasn't able to infer the correct type, so I had to add in a type annotation for `Value`, but now it works!

----

Another question: I tried adding in some unit tests, but they don't seem to run. How do I enable the `web_test` feature?

Also, how do I verify that the documentation I'm writing is correct? Do I just run `cargo doc` and inspect the output?

----

Also, I tried making the `attribute_filter` optional, but now I'm getting this error:

```
error[E0277]: the trait bound `webcore::value::Value: std::convert::From<std::option::Option<&[&str]>>` is not satisfied
  --> src/webapi/mutation_observer.rs:92:64
   |
92 |         let attribute_filter: Value = options.attribute_filter.into();
   |                                                                ^^^^ the trait `std::convert::From<std::option::Option<&[&str]>>` is not implemented for `webcore::value::Value`
   |
   = help: the following implementations were found:
             <webcore::value::Value as std::convert::From<webcore::value::Undefined>>
             <webcore::value::Value as std::convert::From<&'a webcore::value::Undefined>>
             <webcore::value::Value as std::convert::From<&'a mut webcore::value::Undefined>>
             <webcore::value::Value as std::convert::From<webcore::value::Null>>
           and 51 others
   = note: required because of the requirements on the impl of `std::convert::Into<webcore::value::Value>` for `std::option::Option<&[&str]>`
```

Why doesn't `stdweb` define a generic impl that'll work for any `Option`?

**Comment by Pauan** [comment]
> Another question: I tried adding in some unit tests, but they don't seem to run. How do I enable the `web_test` feature?

Ahh, I figured this one out: I was using an older version of `cargo-web`
